Applying Gradient Boosting Regressor to the austin and airbnb dataset, focusing on:

- Gradient Boosting model

- Hyperparameters: learning_rate, n_estimators, max_depth

- Regularization: subsampling (bagging) and min_samples_split


In [4]:
# Gradient Boosting Code (with tuning)

import pandas as pd
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.metrics import mean_squared_error, r2_score
import numpy as np


### Gradient Boosting on Austin Dataset

In [5]:
# Load data
austin_df = pd.read_csv("/Users/sohanaziz/Documents/AI for leaders sem 2/Datasets/cleaned datasets/week 2/cleaned_austin_data.csv")

austin_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15171 entries, 0 to 15170
Data columns (total 44 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   city                            15171 non-null  object 
 1   zipcode                         15171 non-null  int64  
 2   propertyTaxRate                 15171 non-null  float64
 3   garageSpaces                    15171 non-null  int64  
 4   hasAssociation                  15171 non-null  bool   
 5   hasCooling                      15171 non-null  bool   
 6   hasHeating                      15171 non-null  bool   
 7   hasSpa                          15171 non-null  bool   
 8   hasView                         15171 non-null  bool   
 9   parkingSpaces                   15171 non-null  int64  
 10  yearBuilt                       15171 non-null  int64  
 11  latestPrice                     15171 non-null  float64
 12  numPriceChanges                 

In [3]:
# Drop irrelevant columns
austin_df = austin_df.drop(columns=["city", "latest_saledate"])

# Define features and target
X = austin_df.drop(columns=["latestPrice"])
y = austin_df["latestPrice"]

# Train-test split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Initialize model
gbr = GradientBoostingRegressor(random_state=42)

# Define hyperparameter grid
param_grid = {
    'learning_rate': [0.01, 0.05, 0.1],
    'n_estimators': [100, 200],
    'max_depth': [3, 5, 7],
    'subsample': [0.8, 1.0],
    'min_samples_split': [2, 10],
}

# Grid search with 3-fold cross-validation
grid_search = GridSearchCV(
    estimator=gbr,
    param_grid=param_grid,
    cv=3,
    scoring='neg_mean_squared_error',
    n_jobs=-1,
    verbose=1
)

grid_search.fit(X_train, y_train)

# Best model
best_model = grid_search.best_estimator_
y_pred = best_model.predict(X_test)

# Evaluation
rmse = np.sqrt(mean_squared_error(y_test, y_pred))
r2 = r2_score(y_test, y_pred)

# Results
print("\nBest Parameters Found:")
print(grid_search.best_params_)
print(f"Test RMSE: {rmse:.2f}")
print(f"Test R²: {r2:.4f}")

Fitting 3 folds for each of 72 candidates, totalling 216 fits

Best Parameters Found:
{'learning_rate': 0.05, 'max_depth': 5, 'min_samples_split': 2, 'n_estimators': 200, 'subsample': 0.8}
Test RMSE: 314044.83
Test R²: 0.5833


I trained a Gradient Boosting model to predict Austin home prices using a detailed hyperparameter search with 72 configurations and 3-fold cross-validation. The best model achieved an RMSE of $314,045 and explained about 58 percent of the price variance (R² = 0.58) on the test data.

The optimal settings included a learning rate of 0.05, 200 trees, a tree depth of 5, and using 80 percent of data per tree to reduce overfitting. These choices helped balance model accuracy and complexity.

Though the model performs well, adding more detailed features like property condition or neighborhood quality could improve predictions further.

### Gradient Boosting on Airbnb Dataset

In [17]:
# Load Airbnb dataset
airbnb_df = pd.read_csv("/Users/sohanaziz/Documents/AI for leaders sem 2/Datasets/cleaned datasets/week 2/listings_cleaned.csv")

airbnb_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9697 entries, 0 to 9696
Data columns (total 69 columns):
 #   Column                                        Non-Null Count  Dtype  
---  ------                                        --------------  -----  
 0   id                                            9697 non-null   int64  
 1   last_scraped                                  9697 non-null   object 
 2   name                                          9697 non-null   object 
 3   host_id                                       9697 non-null   int64  
 4   host_name                                     9697 non-null   object 
 5   host_since                                    9697 non-null   object 
 6   host_location                                 9697 non-null   object 
 7   host_response_time                            9697 non-null   object 
 8   host_response_rate                            9697 non-null   object 
 9   host_acceptance_rate                          9697 non-null   o

In [18]:
# Drop non-numeric and irrelevant columns
drop_cols = airbnb_df.select_dtypes(include=['object']).columns.tolist() + ['id']
airbnb_df = airbnb_df.drop(columns=drop_cols)

# Drop rows with missing values
airbnb_df = airbnb_df.dropna()

# Define features and target
X = airbnb_df.drop(columns=["price"])
y = airbnb_df["price"]

# Train-test split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Initialize model
gbr = GradientBoostingRegressor(random_state=42)

# Define hyperparameter grid

param_grid = {
    'learning_rate': [0.05, 0.1, 0.15],
    'n_estimators': [300, 350, 400],
    'max_depth': [3, 5, 7],
    'subsample': [0.8, 0.9, 1.0],  
    'min_samples_split': [2, 5],
}  

# Grid search with 3-fold cross-validation
grid_search = GridSearchCV(
    estimator=gbr,
    param_grid=param_grid,
    cv=3,
    scoring='neg_mean_squared_error',
    n_jobs=-1,
    verbose=1
)

grid_search.fit(X_train, y_train)

# Best model
best_model = grid_search.best_estimator_
y_pred = best_model.predict(X_test)

# Evaluation
rmse = np.sqrt(mean_squared_error(y_test, y_pred))
r2 = r2_score(y_test, y_pred)

# Results
print("\nBest Parameters Found:")
print(grid_search.best_params_)
print(f"Test RMSE: {rmse:.2f}")
print(f"Test R²: {r2:.4f}")

Fitting 3 folds for each of 162 candidates, totalling 486 fits

Best Parameters Found:
{'learning_rate': 0.1, 'max_depth': 5, 'min_samples_split': 2, 'n_estimators': 350, 'subsample': 1.0}
Test RMSE: 262.08
Test R²: 0.7201


I trained a Gradient Boosting Regressor on the Airbnb dataset using a wider range of hyperparameters (162 combinations) and 3-fold cross-validation. The best model had a learning rate of 0.1, 350 estimators, max depth of 5, min_samples_split of 2, and subsample of 1.0. It achieved a test RMSE of 262.08 and an R² of 0.7201.

Previously, my best model used 200 estimators with the same values for the other parameters and achieved an RMSE of 261.24 and R² of 0.7219. Since the newer model didn't show significant improvement, it suggests my earlier tuning was already close to optimal. Further parameter expansion brought only minor changes in performance.